Expose path to autoload in a global var for binaries

helhum opened the pull request:
For binaries of packages to be able include the autoload file,
a global variable is now exposed and a proxy file now generated
by default, which exposes the path and includes the original
binary file.

Additionally it is now checked on binary creation whether
the reference binary has a shebang and if not, generates
a much simple proxy code with no magic and eval involved.

Fixes: #10119

* [x] Decide whether we can rely on `vendor-dir` being injected, or what to do when the object is created by other installers not providing it.
* [x] Should symlink creation for binaries remain or be removed?
* [x] Depending on the decision above, maybe add code to migrate from symlinks to proxies, when symlinks are found

helhum commented:
@Seldaek here is my first shot. I will add a few todos, which came to my mind for things to be discussed

Seldaek commented on src/Composer/Installer/BinaryInstaller.php:
```diff
@@ -92,12 +96,14 @@ public function installBinaries(PackageInterface $package, $installPath, $warnOn
             } elseif ($this->binCompat === "full") {
                 $this->installFullBinaries($binPath, $link, $bin, $package);
             } elseif ($this->binCompat === "symlink") {
                 $this->installSymlinkBinaries($binPath, $link);
```
I think I'd treat `symlink` the same as `proxy`, and get rid of `installSymlinkBinaries` altogether, then at least we get some simplification out of this.

helhum replied:
Makes sense. I will remove this code then and replace it with proxy code. I'd suggest to also change the config then to "full" and "proxy" and only keep "symlink" in the config for BC reasons.

Seldaek replied:
Right symlink should just remain supported but silently switch to behave as proxy

helhum replied:
Done

Seldaek commented on src/Composer/Installer/BinaryInstaller.php:
```diff
@@ -237,7 +237,22 @@ protected function generateUnixyProxyCode($bin, $link)
+ *
+ * @generated
+ */
+\$GLOBALS['_composer_autoload_path'] = $autoloadPathCode;$evalCode
```
I wonder if we also want to expose the root dir i.e. `getcwd()` and perhaps simply expose the vendor dir instead of path to autoload.php?

helhum replied:
Hm, not sure. In general I'd say root-dir and vendor-dir *could* be useful during runtime, but I'd rather expose them in `InstalledVersions`. Root dir is btw indirectly exposed already, as the "install path" of the root package equals the root dir.

vendor-dir is indeed missing and could only be indirectly derived by using `dirname(InstalledVersions::getInstallPath('my/package-in-vendor'), 2)`, but I would not expose any of those paths in a global var in the proxy file of binaries.

Seldaek replied:
Alright, let's leave it as is.

Seldaek reviewed (changes requested):
Thanks for the PR.

- This needs a rebase for sure, as the new code is very different (sorry:).
- You should bump the `COMPOSER_RUNTIME_API` to 2.2.0 in `Composer\Composer` as this adds a new project-runtime feature that packages may want to depend on.
- This also raises the question to me of what happens if the vendor-dir changes, and the bin-dir is not inside it, then bin proxies will not get re-generated and they'll suddenly contain incorrect information? We probably should detect/check that and reinstall the proxies in that case? Complexity++ though..

helhum commented:
Thanks @Seldaek for looking into this!

> This also raises the question to me of what happens if the vendor-dir changes, and the bin-dir is not inside it, then bin proxies will not get re-generated and they'll suddenly contain incorrect information? We probably should detect/check that and reinstall the proxies in that case? Complexity++ though..

Hm, isn't that the same with Symlinks? Since Symlinks are relative they will break as well wenn changing bin folder relative to the vendor folder. I'd always considered this acceptable and removed the old bin folder to force a re-install.

> This needs a rebase for sure, as the new code is very different (sorry:).

Will do, no problem!

> You should bump the COMPOSER_RUNTIME_API to 2.2.0 in Composer\Composer as this adds a new project-runtime feature that packages may want to depend on.

Yeah, good idea, will do.

Seldaek commented:
> Hm, isn't that the same with Symlinks? Since Symlinks are relative they will break as well wenn changing bin folder relative to the vendor folder. I'd always considered this acceptable and removed the old bin folder to force a re-install.

I guess you're right :) If we could live with it until now we can probably leave it as is.

helhum commented:
Thanks for the additional fixes @Seldaek! If I can be of any additional help, let me know.

Seldaek commented:
Heh no worries, it's a case of the "the more you look at it the more problems you find" :D 

Wrote some docs too, and fixed more edge cases. I believe it's ready to go now.

Seldaek reviewed (approved):
LGTM now. Happy if anyone else can review though.

helhum commented:
> Heh no worries, it's a case of the "the more you look at it the more problems you find" :D

Yeah. Although I know Composer pretty well, the things you now fixed weren't obvious to me. I will e.g. try to keep in mind that docs are part of this repo too. 👍

> Wrote some docs too, and fixed more edge cases. I believe it's ready to go now.

 Awesome! Thanks again!
